docs/_posts/2022-10-31-luau-semantic-subtyping.md

```diff
@@ -0,0 +1,287 @@
+
+## Subtyping as a source of false positives
+
+One of the sources of false positives in Luau (and many other similar languages like TypeScript or Flow) is *subtyping*. Subtyping is used whenever a variable is initialized or assigned to, and whenever a function is called: the type system checks that the type of the expression is a subtype of the type of the variable. For example, if we add types to the above program
+```lua
+  local x : CFrame = CFrame.new()
+  local y : Vector3 | CFrame
+  if (math.random()) then
+    y = CFrame.new()
+  else
+    y = Vector3.new()
+  end
+  local z : Vector3 | CFrame = x * y
+```
+then the type system checks that the type of `CFrame` multiplication is a subtype of `(CFrame, Vector3 | CFrame) -> (Vector3 | CFrame)`.
+
+Subtyping is a very useful feature, and it supports rich type constructs like type union (`T | U`) and intersection (`T & U`). For example, `number?` is implemented as a union type `(number | nil)`, inhabited by values that are either numbers or `nil`.
+
+Unfortunately, the interaction of subtyping with intersection and union types can have odd results. A simple (but rather artificial) case in older Luau was:
+```lua
+  local x : (number?) & (string?) = nil
+  local y : nil = nil
+  y = x -- Type '(number?) & (string?)' could not be converted into 'nil'
+  x = y
+```
+This error is caused by a failure of subtyping, the old subtyping algorithm reports that `(number?) & (string?)` is not a subtype of `nil`. This is a false positive, since `number & string` is uninhabited, so the only possible inhabitant of `(number?) & (string?)` is `nil`.
+
+This is an artificial example, but there are real issues raised by creators caused by the problems, for example <https://devforum.roblox.com/t/luau-recap-july-2021/1382101/5>. Currently, these issues mostly affect creators making use of sophisticated type system features, but as we make type inference more accurate, union and intersection types will become more common, even in code with no type annotations.
+
+This class of false positives no longer occurs in Luau, as we have moved from our old approach of *syntactic subtyping* to an alternative called *semantic subtyping*.
+
+## Syntactic subtyping
+
+AKA “what we did before.”
+
+Syntactic subtyping is a syntax-directed recursive algorithm. The interesting cases to deal with intersection and union types are:
+
+* Reflexivity: `T` is a subtype of `T`
+* Intersection L: `(T₁ & … & Tⱼ)` is a subtype of `U` whenever some of the `Tᵢ` are subtypes of `U`
+* Union L: `(T₁ | … | Tⱼ)` is a subtype of `U` whenever all of the `Tᵢ` are subtypes of `U`
+* Intersection R: `T` is a subtype of `(U₁ & … & Uⱼ)` whenever `T` is a subtype of some of the `Uᵢ`
+* Union R: `T` is a subtype of `(U₁ | … | Uⱼ)` whenever `T` is a subtype of all of the `Uᵢ`.
```

Comment:
This should be the other way around.

`T` is a subtype of an intersection when it's a subtype of all `Ui`.
`T` is a subtype of a union when it's a subtype of some `Ui`.

Like you have below:
> so by Union R: `nil` is a subtype of `number?`

This is correct, `nil` is only a subtype of `nil` part in the `number?` union, not the `number` and it works out.